File: Volume017/01709 - Amalgamated Artichokes.py
import math
import sys

def price(p, a, b, c, d, k):
    return p*(math.sin(a*k + b) + math.cos(c*k + d) + 2)

def main():
    for line in sys.stdin.readlines():
        p, a, b, c, d, n = [int(i) for i in line.split(' ')]

        max_num = -math.inf
        max_diff = 0

        for i in range(1, n + 1):
            val = price(p, a, b, c, d, i)
            diff = max_num - val

            if diff > max_diff:
                max_diff = diff
            if val > max_num:
                max_num = val

        print(max_diff)

if __name__ == '__main__':
    main()

File: Volume017/01709 - Amalgamated Artichokes.cpp
#include <cstdio>
#include <cmath>

double price(int p, int a, int b, int c, int d, int k) {
    return p*(sin(a*k + b) + cos(c*k + d) + 2);
}

int main() {
    int p, a, b, c, d, n;

    while (scanf("%d %d %d %d %d %d", &p, &a, &b, &c, &d, &n) == 6) {
        double max_num = -1e10;
        double max_diff = 0;

        for (int i = 1; i <= n; i++) {
            double val = price(p, a, b, c, d, i);
            double diff = max_num - val;

            if (diff > max_diff) {
                max_diff = diff;
            }
            if (val > max_num) {
                max_num = val;
            }
        }

        printf("%.6f\n", max_diff);
    }
}

